microbench/src/main/java/io/netty/handler/codec/mqtt/MqttMessageTypeValueOfBench.java

```diff
@@ -0,0 +1,166 @@
+            3, 1, 12, 13, 9, 8, 1, 9, 9, 4, 2, 7, 3, 4, 11, 11,
+            8, 10, 14, 5, 14, 1, 10, 10, 13, 5, 6, 13, 14, 5, 7, 11, 4, 13, 3, 14, 7, 2, 10, 13, 2, 4, 14, 5, 1, 12, 3,
+            13, 11, 2, 11, 14, 2, 5, 8, 13, 4, 13, 13, 3, 3,
+            3, 13, 6, 11, 5, 3, 2, 13, 9, 2, 10, 8, 3, 11, 4, 6, 12, 14, 6, 2, 14, 1, 2, 6, 8, 4, 12, 8, 11, 9, 1, 7, 1,
+            10, 4, 10, 9, 9, 3, 11, 5, 10, 8, 9, 4, 13, 4,
+            5, 7, 12, 14, 12, 6, 1, 2, 10, 9, 10, 12, 1, 2, 6, 9, 5, 13, 4, 6, 11, 7, 1, 3, 10, 2, 1, 13, 14, 3, 5, 5,
+            5, 7, 14, 9, 9, 3, 12, 1, 1, 1, 3, 12, 6, 9, 7, 8, 1,
+            8, 2, 8, 13, 1, 11, 11, 1, 4, 10, 4, 3, 10, 3, 2, 2, 8, 2, 4, 13, 14, 4, 12, 14, 7, 6, 7, 13, 7, 11, 13, 12,
+            14, 1, 14, 3, 4, 13, 12, 10, 5, 12, 12, 4, 5, 6,
+            9, 12, 13, 3, 4, 13, 8, 14, 3, 2, 8, 5, 6, 13, 8, 7, 4, 5, 8, 14, 8, 14, 7, 5, 4, 9, 12, 12, 10, 3, 1, 12,
+            5, 1, 11, 6, 10, 5, 14, 4, 5, 13, 8, 11, 13, 4, 9,
+            9, 7, 6, 2, 2, 5, 12, 13, 13, 6, 11, 13, 12, 10, 6, 7, 1, 2, 6, 1, 9, 10, 14, 7, 9, 2, 2, 2, 8, 8, 11, 14,
+            12, 9, 13, 1
+    };
+
+    int[] types;
+
+    @Setup
+    public void initDataSet() {
+        this.types = DATASET;
+    }
+
+    @Benchmark
+    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
+    public void getViaArray(Blackhole bh) {
+        for (int type : types) {
+            bh.consume(MqttMessageTypeArray.valueOf(type));
+        }
+    }
+
+    @Benchmark
+    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
+    public void getViaSwitch(Blackhole bh) {
+        for (int type : types) {
+            bh.consume(switchValueOf(type));
+        }
+    }
+
+    private static MqttMessageType switchValueOf(int type) {
+        switch (type) {
+        case 1:
+            return MqttMessageType.CONNECT;
+        case 2:
+            return MqttMessageType.CONNACK;
+        case 3:
+            return MqttMessageType.PUBLISH;
+        case 4:
+            return MqttMessageType.PUBACK;
+        case 5:
+            return MqttMessageType.PUBREC;
+        case 6:
+            return MqttMessageType.PUBREL;
+        case 7:
+            return MqttMessageType.PUBCOMP;
+        case 8:
+            return MqttMessageType.SUBSCRIBE;
+        case 9:
+            return MqttMessageType.SUBACK;
+        case 10:
+            return MqttMessageType.UNSUBSCRIBE;
+        case 11:
+            return MqttMessageType.UNSUBACK;
+        case 12:
+            return MqttMessageType.PINGREQ;
+        case 13:
+            return MqttMessageType.PINGRESP;
+        case 14:
+            return MqttMessageType.DISCONNECT;
+        default:
+            throw new IllegalArgumentException("unknown message type: " + type);
+        }
+    }
+
+    private static class MqttMessageTypeArray {
+        private static final MqttMessageType[] VALUES;
```

Comment:
Not sure we need to keep this copy around if we're decided on going with the array?

Reply:
we won't indeed, good point!